src/main/ts/core/Changelog.ts

```diff
@@ -0,0 +1,361 @@
+}
+
+export interface Changelog {
+  source: string;
+  releases: Release[];
+  preamble: Offset;
+  links: Offset;
+}
+
+interface TextLines {
+  text: string;
+  lines: number[];
+}
+
+interface Heading {
+  header: commonmark.Node;
+  preamble: commonmark.Node[];
+  subheadings: Heading[];
+}
+
+interface Top {
+  header?: undefined;
+  preamble: commonmark.Node[];
+  subheadings: Heading[];
+}
+
+const findLineStarts = (text: string): TextLines => {
+  const lines = [ 0 ];
+  let seenCR = false;
+  let seenLF = false;
+  for (let i = 0; i < text.length; i++) {
+    const c = text.charAt(i);
+    if (c === '\n') {
+      if (seenLF) {
+        // add line start and reset
+        lines.push(i);
+        seenCR = false;
+        seenLF = false;
+      }
+      seenLF = true;
+    } else if (c === '\r') {
+      if (seenCR) {
+        // add line start and reset
+        lines.push(i);
+        seenCR = false;
+        seenLF = false;
+      }
+      seenCR = true;
+    } else {
+      if (seenLF || seenCR) {
+        // add line start and reset
+        lines.push(i);
+        seenLF = false;
+        seenCR = false;
+      }
+    }
+  }
+  return { text, lines };
+};
+
+const posToOffset = (lines: number[], position: [number, number]) => {
+  // both line and column count from 1
+  const [ line, column ] = position;
+  return lines[line - 1] + (column - 1);
+};
+
+// the range including just the columns of the node
+const columnRange = (lines: number[], node: commonmark.Node): Offset => {
+  // this is the inclusive range
+  const [ startPos, endPos ] = node.sourcepos;
+  const start = posToOffset(lines, startPos);
+  const end = posToOffset(lines, endPos);
+  return { start, end };
+};
+
+// the range including the whole lines that the node is in
+const blockRange = (source: TextLines, node: commonmark.Node): Offset => {
+  const [[ startLine ], [ endLine ]] = node.sourcepos;
+  const start = posToOffset(source.lines, [ startLine, 1 ]);
+  const end = (endLine < source.lines.length ? posToOffset(source.lines, [ endLine + 1, 1 ]) : source.text.length) - 1;
+  return { start, end };
+};
+
+const extractText = (source: TextLines, node: commonmark.Node): string => {
+  const range = columnRange(source.lines, node);
+  return source.text.substring(range.start, range.end + 1);
+};
+
+const pos = (node: commonmark.Node): string => {
+  if (node.sourcepos !== undefined) {
+    const [[ startLine, startColumn ]] = node.sourcepos;
+    return ` (line: ${startLine} column: ${startColumn})`;
+  } else {
+    return '';
+  }
+};
+
+const parseHeadings = (changelog: string): Top => {
+  const data: Top = { preamble: [], subheadings: [] };
+  const reader = new commonmark.Parser();
+  const doc = reader.parse(changelog);
+  const stack: (Heading | Top)[] = [ data ];
+  let node: commonmark.Node | null = doc.firstChild;
+  while (node !== null) {
+    if (node.type === 'heading') {
+      // pop headings off the stack that are higher or equal level
```

Comment:
I'm not sure if it needs to happen here or above where the stack is declared, but can you expand on this to explain why we want to do that?

From staring at the code a while, I was able to piece together that we don't need to know about the latest `h3` and `h4` (or the latest `h2`) any more once we find a new `h2`, as no new content is going to become a child of those old headings. But it wasn't immediately obvious from the code

Reply:
It is as you surmised. 

There are 2 senariios:
1. We find a header and it is a larger number (ie we were parsing a h2 and we see a h3) then the new heading is a subheading of the old heading.
2. We find a header and it is a smaller number (ie we were parsing a h3 and we see a h2) then the new heading is not a sub heading of the current one. We need to find the next heading that is a smaller number in the stack that we've seen (ie we pop off the h3) and then the new heading is the subheading of the last item in the stack not popped.

Reply:
If the question is why not use an array and then just take the index before the current as the owner, well it is possible to have a h3 as the child of a h1 if there are no h2 in the document so the stack worked better.

Reply:
I've added an explanation to the code and renamed a few variables to hopefully make this clearer.